### ISSUE TITLE
fused-engine: generation stays degenerate (always token 0) on every --policy — O-proj/QKV/FFN unconditionally routed through an unimplemented NPU worker stub

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#56]
## Summary

Following the #44 fix (real I8 dequantization for `lm_head`/`embed_tokens`, landed in 85d907d2) and a follow-up fix for corrupted NaN-pattern scale bytes found in the real `lm_head.weight` tensor (501d0e43), end-to-end generation with `fused-engine` is **still** completely degenerate — it outputs token 0 repeated for every position, regardless of `--policy npu_only|gpu_only|cpu_only`. Root-caused to two compounding bugs:

## Bug 1: dispatch policy is not consulted for O-projection anywhere, and only partially consulted for QKV/FFN

`self.npu.runOProj(...)` is called completely unconditionally — with a plain `try`, no dispatch check at all — in every single code path that computes attention output:

- `prefill()` (`engine/fusion/fused_execute.zig:2168`)
- `forwardDecodeBatchGEMV()` (`fused_execute.zig:1872`) — **this is the path actually exercised by default**, since `BATCH_SIZE = 128` (`main.zig:21`) always exceeds `BATCH_SPLIT_THRESHOLD = 5` (`fused_execute.zig:91`), so `forwardDecode` immediately redirects to the batched-GEMV path (`fused_execute.zig:1940`)
- `executeLayerAttnFFN()` (`fused_execute.zig:1514`) — the "skinny" per-token path, only reached when batch ≤ 5

QKV projection (`runQKV`) and FFN gate/up/down (`runFFN`/`runDown`) *are* gated by `dispatch.qkv`/`dispatch.ffn` in the skinny path (`executeLayerQKV`, `runDenseFfn`'s caller) — but **not** in `prefill()` or `forwardDecodeBatchGEMV()`, which call them unconditionally too. Since the default batch size never reaches the skinny path, in practice **none** of QKV/O-proj/FFN respect `--policy` at all during normal operation.

## Bug 2: the NPU worker subprocess is an unimplemented stub

`engine/npu/src/npu_engine_universal.cpp:286`, inside `--worker` mode (the persistent subprocess `FusedExecutor` spawns for every policy):

```cpp
std::vector<float> out_data(batch*out_dim,0);
// For now: zero output (no NPU hardware access in worker mode)
// The fused engine will use CPU fallback
uint32_t resp[2]={0,out_dim};
```

Every QKV/O-proj/gate-up/down request over the worker protocol returns all-zero output, unconditionally, regardless of input.

## Combined effect

Since attention (O-proj) and FFN both always route through `self.npu.*`, and that always returns zero, their contribution to the residual stream is exactly zero at every layer, for every policy. The 28-layer transformer body collapses to pure token-embedding pass-through: `hidden_final ≈ embed_tokens[input_token]`. `lmHead()` then computes a real (now that #44 and the NaN-sanitization fix are in) but architecturally meaningless argmax over `dot(embedding, lm_head_row)` — which converges to a fixed point (repeatedly re-embedding whichever token wins) rather than producing coherent text.

## Repro

```bash
cd engine/fusion && zig build
./zig-out/bin/fused-engine -m /path/to/model.q4nx -p "The capital of France is" -n 15 --policy cpu_only
# (same result with --policy gpu_only or --policy npu_only)
```

Output: `0 0 0 0 0 0 0 0 0 0 0 0 0 0 0` — regardless of which policy is passed. Verified with debug instrumentation that `hidden` going into `lmHead()` is real/finite/nonzero (matches the raw input-token embedding almost exactly across decode steps), and that logits are now finite (post the NaN-sanitization fix) rather than inf/-inf.

Also: the CLI process reliably dies with `SIGABRT` (exit 134, core dumped) right after printing the generation summary, preceded by `[Vulkan Loader] ERROR: vkDeviceWaitIdle: Invalid device`. Distinct from #53 (that one is `npu_engine_all`'s `free(): invalid size`; this is `fused-engine` itself, at Vulkan/GPU teardown). Not root-caused further here.

## What's needed to actually fix this (not attempted — scoped investigation only)

Checked whether an existing CPU or GPU matmul path could just be wired in instead of the NPU stub:

- **CPU**: `cpu_backend.zig` + `cpu_layer.cpp` have a complete, real CPU transformer-layer implementation (`cpu_layer_forward_qwen3`) — but it operates on a **2-bit packed ternary weight format**, incompatible with the I8/Q4NX 4-bit tile format `model.q4nx` actually uses. `CpuBackend.loadWeights()` is `return error.NotImplemented`, and `fused_execute.zig` never imports/instantiates `CpuBackend` at all — the module's own header comment is a TODO for wiring it in that was never done.
- **GPU**: `gpu_attn.zig` only implements flash attention — no GEMV/GEMM shader for linear projections exists in this engine. The separate `zinc` project has a large dequant-matmul shader library (`dmmv_q4k_*`, `mul_mm_q5k_*`, etc.) but none target the Q4NX tile format; only `flash_attn_batched.spv` is currently borrowed from it.
- `dequant_i8_to_float_ex` (`engine/npu/src/dequant_q4nx.c`) is compiled and linked into the `fused-engine` binary (`build.zig:32`) but never called from any `.zig` file — dead code, presumably meant for exactly this purpose.

So closing this needs either (a) real NPU worker-mode hardware dispatch, or (b) new CPU dequant+GEMV code for Q4NX weights (the embeddings/lm_head dequantizer from #44 could be extended to per-layer q/k/v/o/gate/up/down weights), or (c) a new Vulkan GEMV shader for the Q4NX format on the GPU side. None of these are wiring fixes — they're new implementation work.

## Related
- #44 — lm_head/embeddings zero-fill (fixed; necessary but, per its own text, not sufficient — this issue is the "not sufficient" part)
- #54 — process gap around hardware validation before merge (this bug is exactly the kind of thing that gap lets through)
- #48, #53 — other NPU engine reliability bugs found in the same family of testing

**Comment by bong-water-water-bong** [comment]
Triage (keeping open): this is the umbrella broken-generation bug and needs on-device (gfx1151 + XDNA2) validation I can't do from here. Related fixes landed that reduce the ways it hides: #145 (silent `@memset(0)` mis-dispatch fallbacks now `log.warn` loudly), #113/#144 (teardown double-free/leak fixed), #142 (decode-token overflow cap). The core work — real NPU worker-mode dispatch, or a Q4NX CPU/GPU GEMV path — remains. Per the latest hardware note (commit `d4ac78b0`) the current blocker is `DRM_IOCTL_AMDXDNA_CREATE_HWCTX` EINVAL at the resource-solver layer, not xclbin column targeting.

**Comment by bong-water-water-bong** [comment]
Not fixed in #177 — flagging why. This needs the NPU to validate any fix (the worker stub returns zeros because the on-NPU GEMM path isn't wired end-to-end, and I can't build/run the XRT + XDNA path on a box without the hardware attached).

Context from the #177 work: the CPU-path degenerate output often conflated with this stub had a **separate, now-fixed root cause** (#153 packed-nibble index). Once #177 lands, please re-test `--policy cpu_only` first to confirm it now produces coherent tokens — that isolates remaining degeneracy to the NPU worker stub (this issue) rather than the dequant bug (fixed).
